pandas/core/indexes/api.py

```diff
@@ -243,7 +244,30 @@ def conv(i):
                 i = i.tolist()
             return i
 
-        return Index(lib.fast_unique_multiple_list([conv(i) for i in inds], sort=sort))
+        return Index(
+            lib.fast_unique_multiple_list([conv(i) for i in inds], sort=sort),
+            dtype=dtype,
+        )
+
+    def _find_common_index_dtype(inds):
+        """
+        Finds a common type for the indexes to pass through to resulting index.
+
+        Parameters
+        ----------
+        inds: list of Index or list objects
+
+        Returns
+        -------
+        The common type or None if no indexes were given
+        """
+        dtypes = [idx.dtype for idx in indexes if isinstance(idx, Index)]
+        if dtypes:
+            dtype = find_common_type(dtypes)
```

Comment:
I assume we never pass a mixed list of Indexes and lists? could add type annotations here for clarity.

Reply:
A parent function uses list[list[Hashable] | Index], so not sure

Reply:
I assumed that if we could pass a mixed list, then the logic added here would not account for the types in a list and only use the Indexes to find the common dtype. We could therefore expect this to raise in those cases?

Reply:
> A parent function uses `list[list[Hashable] | Index]`, so not sure

we may need to change to `list[list[Hashable]] | list[Index]]`